File: exercises/reto3.py
"""
String Challenge

Have the function

StringChallenge (str) take the str
parameter being passed and modify it
using the following algorithm. Replace
every letter in the string with the letter
following it in the alphabet (ie. c
becomes d, z becomes a). Then
capitalize every vowel in this new string
(a, e, i, o, u) and finally return this
modified string.

Once your function is working, take
the final output string and intersperse

    it character-by-character with your
    ChallengeToken.

    Your ChallengeToken: hbs2oim9c1a

Examples

    Input: "hello*3"
    Output: Ifmmp*3
    Final Output: Ihfbmsm2po*i3m9c1a

    Input: "fun times!"
    Output: gvO Ujnft!
    Final Output: ghvbOs 2Uojinmf9tc!1a
"""

from itertools import zip_longest


def StringChallenge(str):
    dic = {
        "a": "b",
        "b": "c",
        "c": "d",
        "d": "e",
        "e": "f",
        "f": "g",
        "g": "h",
        "h": "i",
        "i": "j",
        "j": "k",
        "k": "l",
        "l": "m",
        "m": "n",
        "n": "o",
        # "ñ": "o",
        "o": "p",
        "p": "q",
        "q": "r",
        "r": "s",
        "s": "t",
        "t": "u",
        "u": "v",
        "v": "w",
        "w": "x",
        "x": "y",
        "y": "z",
        "z": "a",
    }
    cadena = ""
    token = "hbs2oim9c1a"
    vocales = "aeiou"

    for c in str:
        if c in dic.keys():
            if dic[c] in vocales:
                cadena += dic[c].upper()
            else:
                cadena += dic[c]
        else:
            cadena += c

    resultado = zip_longest(cadena, token, fillvalue="")

    return "".join(a + b for a, b in resultado)


r1 = StringChallenge("hello*3")
r2 = StringChallenge("fun times!")
print(r1 + "\n" + r2)

File: exercises/reto3.java
import java.util.HashMap;
import java.util.Map;

public class StringChallenge {
    
    public static String stringChallenge(String str) {
        // Crear el diccionario de mapeo
        Map<Character, Character> dic = new HashMap<>();
        String alphabet = "abcdefghijklmnopqrstuvwxyz";
        
        for (int i = 0; i < alphabet.length(); i++) {
            char current = alphabet.charAt(i);
            char next = alphabet.charAt((i + 1) % alphabet.length());
            dic.put(current, next);
        }
        
        StringBuilder resultado = new StringBuilder();
        String token = "hbs2oim9c1a";
        String vocales = "aeiou";
        
        // Transformar la cadena
        for (int i = 0; i < str.length(); i++) {
            char c = str.charAt(i);
            char lower = Character.toLowerCase(c);
            
            if (dic.containsKey(lower)) {
                char nuevo = dic.get(lower);
                if (vocales.indexOf(nuevo) != -1) {
                    resultado.append(Character.toUpperCase(nuevo));
                } else {
                    resultado.append(nuevo);
                }
            } else {
                resultado.append(c);
            }
        }
        
        String cadena = resultado.toString();
        
        // Once your function is working, take the final output string and concatenate it with your ChallengeToken, and then replace every fourth character with an underscore.
        StringBuilder finalResult = new StringBuilder();
        String combined = cadena + token;
        for (int i = 0; i < combined.length(); i++) {
            if ((i + 1) % 4 == 0) {
                finalResult.append('_');
            } else {
                finalResult.append(combined.charAt(i));
            }
        }
        
        return finalResult.toString();
    }
    
    public static void main(String[] args) {
        String r1 = stringChallenge("hello*3");
        String r2 = stringChallenge("fun times!");
        System.out.println(r1);
        System.out.println(r2);
    }
}